### PR TITLE
ARROW-11446: [DataFusion] Added support for scalarValue in Builtin functions.

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR adds support for `ScalarValue` to all builtin-functions and UDFs from DataFusion.

This allows to execute builtin functions without having to create arrays and pass then to the kernels.

With this change, coding a new kernel and UDFs is a more time consuming, as it is necessary to cater for both the scalar and array case. OTOH, this leads to a major performance improvement as we only need to perform 1 operation instead of the usual number of rows operations.

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
cc  @alamb @andygrove @Dandandan @seddonm1

**Comment by seddonm1** [comment]
@jorgecarleitao yes this looks good 👍 

This is going to take some time to apply but I like the idea of passing a function to apply to the string (for example). It would be good to get this merged soon so I can apply to the big Postgres functions PR.

**Comment by jorgecarleitao** [comment]
@alamb , when you wrote the `date_trunc`, did you had in mind having different granularities per array, or was it driven by the fact the the builtin functions only accepted arrays?

If the latter, do you think it would be ok to require granularity to always be a scalar as a regression / feature of this PR?

I am asking because `date_trunc(array | scalar, array | scalar)` is a "bit" more effort than `date_trunc(scalar, array | scalar)`.

**Comment by Dandandan** [comment]
FWIW, PostgreSQL also doesn't support the non-scalar granularity for `date_trunc`

**Review by alamb — COMMENTED** [review]
I like where this is headed @jorgecarleitao ❤️ I skimmed it quickly but will be happy to review it thoroughly when it is ready

Another thing this PR's functionality is likely to enable is more general 'constant folding' -- e.g the kind of thing @houqp  is doing for booleans in https://github.com/apache/arrow/pull/9309

Imagine an expression like `A < (5 + 1)` -- using this code we could evalaute the `5+1` into `6` at PLAN time rather than during execution. A more likely useful example is `date_truc(date_col, 'month') < date_trun(2021-01-31, 'month')` and other similar things

**Comment by alamb** [comment]
@jorgecarleitao  -- I agree with @Dandandan  -- think `date_trunc(scalar, array)`  and `date_trun(scalar, scalar) are the really important cases. I have no use case for `date_trunc(array, array)`

**Comment by jorgecarleitao** [comment]
This is ready for a first take.

Some notes:

* This changes the UDF's API: they should now handle `ColumnarValue`, not `ArrayRef`.
* whenever possible, I used generics instead of macros. I have been reasoning about code more easily with generics because they set the traits explicitly and are therefore IMO easier to use.
* I had to add a new trait to bridge `ScalarValue` and `ArrowPrimitiveType`
* I had to implement the `TryFrom` for the timestampNanoseconds.
* I have not added test coverage to the scalar cases of our existing functions. Longer term, I think we should develop a generic to handle this without having to test the two cases. I tried to do something like this, but I still had to copy-paste some code between generics, unfortunately.

Sorry for the long PR, but I did not find an easy way to split it up (besides #9378, which this PR is built on top of).

**Comment by seddonm1** [comment]
> @jorgecarleitao -- I agree with @Dandandan -- think `date_trunc(scalar, array)` and `date_trun(scalar, scalar) are the really important cases. I have no use case for `date_trunc(array, array)`

Having been trying to implement a lot of the Postgres logic recently they all logically support the use of arrays for each of the parameters (i.e. `date_trunc(array, array)`) I have no idea how common the use of this functionality really is - and I suspect it is low - but if we are building the API for Postgres compatibility it would be good to solve.

**Comment by alamb** [comment]
> Having been trying to implement a lot of the Postgres logic recently they all logically support the use of arrays for each of the parameters (i.e. date_trunc(array, array)) I have no idea how common the use of this functionality really is - and I suspect it is low - but if we are building the API for Postgres compatibility it would be good to solve.

This is a good point @seddonm1 - I am just starting to look at this pR, but I wonder if we could do something like have a default implementation for `func(scalar, scalar)`, `func(scalar, array)`, and `func(array, scalar)` that are implemented in terms of the `func(array, array)`

that way we would only have to supply `func(array, array)` and the scalar input case could be handled on a case by case basis. I'll try and commend inline if I see how that might be done

**Comment by seddonm1** [comment]
@alamb here is a proof-of-concept for the regexp_replace Postgres function which has been built to support the possibility of passing in different function parameters for each row. If it were possible to tell whether the value was a Scalar or Array there would be major optimisation opportunities. I did some basic memoization of the Regex objects but that would not be as necessary if we knew Scalar v Array.

https://github.com/apache/arrow/pull/9243/files#diff-abe8768fe7124198cca7a84ad7b2c678b3cc8e5de3d1bc867d498536a2fdddc7R542

File: rust/datafusion/src/physical_plan/datetime_expressions.rs [hunk 1]
```diff
@@ -167,152 +175,166 @@ fn naive_datetime_to_timestamp(s: &str, datetime: NaiveDateTime) -> Result<i64>
...
-    let result = (0..num_rows)
-        .map(|i| {
-            if string_args.is_null(i) {
-                // NB: Since we use the same null bitset as the input,
-                // the output for this value will be ignored, but we
-                // need some value in the array we are building.
-                Ok(0)
-            } else {
-                string_to_timestamp_nanos(string_args.value(i))
+pub(crate) fn unary_string_to_primitive_function<'a, T, O, F>(
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/datetime_expressions.rs:178]: I think adding an example of using this pattern for implementing UDfs might be really helpful

File: rust/datafusion/src/physical_plan/datetime_expressions.rs [hunk 2]
```diff
@@ -167,152 +175,166 @@ fn naive_datetime_to_timestamp(s: &str, datetime: NaiveDateTime) -> Result<i64>
...
+    let array = args[0]
+        .as_any()
+        .downcast_ref::<GenericStringArray<T>>()
+        .unwrap();
+
+    // first map is the iterator, second is for the `Option<_>`
+    array.iter().map(|x| x.map(|x| op(x)).transpose()).collect()
+}
+
+fn handle<'a, O, F, S>(
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/datetime_expressions.rs:205]: I may be missing it -- but this function only seems to be invoked once with the same set of type arguments -- does it need to be generic? Or more broadly, can we hoist this seemingly repeated `handle` pattern somewhere it can be  reused (and reduce the cognitive load on people writing new functions?)

File: rust/datafusion/src/physical_plan/string_expressions.rs [hunk 3]
```diff
@@ -17,27 +17,100 @@
...
-        $ARGS
-            .iter()
-            .map(|e| match e.as_any().downcast_ref::<$ARRAY_TYPE>() {
-                Some(array) => Ok(array),
-                _ => Err(DataFusionError::Internal("failed to downcast".to_string())),
-            })
-    }};
+use super::ColumnarValue;
+
+pub(crate) fn unary_string_function<'a, T, O, F, R>(
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/string_expressions.rs:33]: this function is effectively applying a `str->str` function `op` to all values in `args`? Some comments might be helpful

**Review by alamb — COMMENTED** [review]
So I wrote up a big comment that github seems to have lost :(

Basically I spent 30 minutes studying this PR. I think it is one of the most important PRs for expression evaluation I have seen in DataFusion.

My major concern (other than the annoyance of the backward compatibility) is that it might put the implementation of built in functions even further out of reach for new contributors. The cognitive load of understanding `ArrayRef` in general and working with them is already large, and having to handle ColumnValues will just make it more so.

Also this PR is too large for me to effectively review -- even with some more time (which I am not sure I can find) I am not sure I could fully grok it. 

I wonder if we could do something like the following to both ease new developers as well as allow this PR to be implemented in chunks.

Make a new type that has the old array signature:

```rust
pub type SimpleScalarFunctionImplementation =
    Arc<dyn Fn(&[ArrayRef]) -> Result<ArrayRef> + Send + Sync>;
```

And then make an adapter function that takes a simple signature and implements the more general one:

```rust
fn make_scalar_function(inner: SimpleScalarFunctionImplementation) -> ScalarFunctionImplementation {
   // invoke inner correctly, creating `ArrayRef` to hold a bunch of scalar values
  }
```

That way you could keep most signatures the same and then implement the general cases one by one. 

Thanks again @jorgecarleitao  -- this is really awesome work,

**Comment by jorgecarleitao** [comment]
Hi @alamb , thanks a lot for your review and comments. I generally agree that this makes it more difficult to write a UDF and a function implementation in general.

I like that idea. I have now pushed a commit with it. I will now address the remaining ones.

File: rust/arrow/src/conversions.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (jorgecarleitao) [reply rust/arrow/src/conversions.rs]: It was renamed on a different PR to `temporal_conversions`. I rebased this PR so that we no longer have these changes.

File: rust/datafusion/examples/simple_udf.rs [hunk 5]
```diff
@@ -54,50 +58,76 @@ fn create_context() -> Result<ExecutionContext> {
...
+    assert_eq!(exponent.len(), base.len());
+
+    // 2. perform the computation
+    let array = base
+        .iter()
+        .zip(exponent.iter())
+        .map(|(base, exponent)| maybe_pow(&base, &exponent))
+        .collect::<Float64Array>();
+
+    // `Ok` because no error occurred during the calculation (we should add one if exponent was [0, 1[ and the base < 0 because that panics!)
```
[the remark this answers is not in the document]
Comment (jorgecarleitao) [reply rust/datafusion/examples/simple_udf.rs]: This represents an open interval: `0 <= exponent < 1` because `(x)^1 = x` for all x (including negative ones).

File: rust/datafusion/src/physical_plan/datetime_expressions.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/datetime_expressions.rs:205]: The pattern depends on the input and output of the function. I.e. when input is `&str`, then `Utf8/LargeUtf8`. When output is a `Native`, then the output is `PrimitiveArray<O::Native>`. In general this construct depends on the what the user is trying to achieve (wrt to input and output types).

I placed this here because it allows to decouple the pattern (of handling Scalar and Array) from the implementation of the logic (`string_to_timestamp_nanos` in this case).

In crypto_expressions we have a similar pattern, but in this case the function is `&str -> AsRef<[u8]>`, which allowed to write all cripto `sha` in a succinct manner. However, in that case, the output type is always `Binary` instead of `LargeBinary` for `LargeStringArray`, because the hashes are always smaller than `i32::MAX`.

**Comment by jorgecarleitao** [comment]
Ok, I have addressed the comments.

The API change to UDFs is that people need to call `make_scalar_function` on their existing UDFs, as seen in diff of the example.

Out of curiosity, did anyone run the benchmarks? I do not have a machine suitable for that, but I am obviously curious :)

File: rust/datafusion/src/physical_plan/datetime_expressions.rs (hunk 2 above)
Comment (alamb) [reply rust/datafusion/src/physical_plan/datetime_expressions.rs:218]: makes sense -- thank you for the explination

File: rust/datafusion/examples/simple_udf.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alamb) [reply rust/datafusion/examples/simple_udf.rs]: cool -- sorry I guess I am used to seeing an open interval  using a `)` --so in this case something like `[0, 1)` to represent `0 <= exponent < 1` (e.g. [here](https://en.wikipedia.org/wiki/Interval_(mathematics)#Terminology)

File: rust/datafusion/examples/simple_udf.rs [hunk 6]
```diff
@@ -58,77 +54,54 @@ fn create_context() -> Result<ExecutionContext> {
...
+                }
+            })
+            .collect::<Float64Array>();
+
+        // `Ok` because no error occurred during the calculation (we should add one if exponent was [0, 1[ and the base < 0 because that panics!)
+        // `Arc` because arrays are immutable, thread-safe, trait objects.
+        Ok(Arc::new(array) as ArrayRef)
+    };
+    // the function above expects an `ArrayRef`, but DataFusion may pass a scalar to a UDF.
+    // thus, we use `make_scalar_function` to decorare the closure so that it can handle both Arrays and Scalar values.
```
Comment (alamb) [inline rust/datafusion/examples/simple_udf.rs:103]: 👍

**Review by alamb — APPROVED** [review]
I think it is looking good now @jorgecarleitao  -- thank you 👍

**Comment by Dandandan** [comment]
@jorgecarleitao what kind of benchmarks are you interested in? AFAIK, most benchmarks are not very depending on this, I expect it being mostly impactfully in cases where the projection itself is a lot of the time, but most benchmarks are spending most of the time on joins/aggregates.

**Comment by alamb** [comment]
@andygrove  / @Dandandan / @seddonm1  / @maxburke / @jhorstmann  -- what are your thoughts on this one?

It is a significant enough change I think someone using DataFusion in their projects, other than myself should weigh in. I like it a lot -- and I think it could serve as the basis for constant folding (e.g. transform `A + (5 + 7)` --> `A + 12` at plan time)

**Comment by Dandandan** [comment]
I think changes here look good, and will be helpful to implement functions which typically operate on `(scalar, array)`. And this will help to make those cases (quite a bit) more efficient. I think that can be particularly helpful when having complex expressions in `WHERE` clauses / projections.

@alamb I am not sure I understand why folding should be done on the physical plan level or really depending on this PR , it should be possible without changes in this PR (just by having some `Expr -> Expr`  rules)? They could share the same evaluation code though?

File: rust/datafusion/src/physical_plan/functions.rs [hunk 7]
```diff
@@ -622,21 +560,66 @@ impl PhysicalExpr for ScalarFunctionExpr {
...
+            args.iter()
+                .map(|arg| arg.clone().into_array(len))
+                .collect::<Vec<ArrayRef>>()
+        } else {
+            args.iter()
+                .map(|arg| arg.clone().into_array(1))
+                .collect::<Vec<ArrayRef>>()
+        };
+
+        let result = (inner)(&args);
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/functions.rs:599]:
```suggestion
        let result = inner(&args);
```

File: rust/datafusion/src/physical_plan/crypto_expressions.rs [hunk 8]
```diff
@@ -49,58 +58,142 @@ fn sha_process<D: SHA2Digest + Default>(input: &str) -> SHA2DigestOutput<D> {
...
+                )?)))
+            }
+            other => Err(DataFusionError::Internal(format!(
+                "Unsupported data type {:?} for function {}",
+                other, name,
+            ))),
+        },
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(a) => {
+                let result = a.as_ref().map(|x| (op)(x).as_ref().to_vec());
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/crypto_expressions.rs:126]:
```suggestion
                let result = a.as_ref().map(|x| op(x).as_ref().to_vec());
```

File: rust/datafusion/src/physical_plan/crypto_expressions.rs [hunk 9]
```diff
@@ -49,58 +58,142 @@ fn sha_process<D: SHA2Digest + Default>(input: &str) -> SHA2DigestOutput<D> {
...
+                Ok(ColumnarValue::Scalar(ScalarValue::Binary(result)))
+            }
+            ScalarValue::LargeUtf8(a) => {
+                let result = a.as_ref().map(|x| (op)(x).as_ref().to_vec());
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/crypto_expressions.rs:130]:
```suggestion
                let result = a.as_ref().map(|x| op(x).as_ref().to_vec());
```

File: rust/datafusion/src/physical_plan/crypto_expressions.rs [hunk 10]
```diff
@@ -49,58 +58,142 @@ fn sha_process<D: SHA2Digest + Default>(input: &str) -> SHA2DigestOutput<D> {
...
+                ])?)))
             }
+            other => Err(DataFusionError::Internal(format!(
+                "Unsupported data type {:?} for function md5",
+                other,
+            ))),
+        },
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(a) => {
+                let result = a.as_ref().map(|x| md5_process(x));
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/crypto_expressions.rs:166]:
```suggestion
                let result = a.as_ref().map(md5_process);
```

**Comment by alamb** [comment]
> @alamb I am not sure I understand why folding should be done on the physical plan level or really depending on this PR , it should be possible without changes in this PR (just by having some Expr -> Expr rules)? They could share the same evaluation code though?

@Dandandan  -- yes, I think constant folding looks like `Expr` --> `Expr`. But as you hint at, if the physical evaluation of functions is entirely separate from the plan time evaluation, we will end up with two parallel implementations for evaluation that need to be kept in sync (one for `Expr` and one for the physical runtime) -- using the same implementation for both I think avoids a lot of potential inconsistency (and source of bugs)

We'll see if that turns out to be possible, but I think it should be

File: rust/datafusion/src/physical_plan/string_expressions.rs [hunk 11]
```diff
@@ -17,27 +17,108 @@
...
+
+    let array = args[0]
+        .as_any()
+        .downcast_ref::<GenericStringArray<T>>()
+        .ok_or_else(|| {
+            DataFusionError::Internal("failed to downcast to string".to_string())
+        })?;
+
+    // first map is the iterator, second is for the `Option<_>`
+    Ok(array.iter().map(|x| x.map(|x| op(x))).collect())
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/string_expressions.rs:66]:
```suggestion
    Ok(array.iter().map(|x| x.map(op).collect())
```

File: rust/datafusion/src/physical_plan/crypto_expressions.rs [hunk 12]
```diff
@@ -49,58 +58,142 @@ fn sha_process<D: SHA2Digest + Default>(input: &str) -> SHA2DigestOutput<D> {
...
+        .ok_or_else(|| {
+            DataFusionError::Internal("failed to downcast to string".to_string())
+        })?;
 
-            let array = args[0]
-                .as_any()
-                .downcast_ref::<GenericStringArray<T>>()
-                .unwrap();
+    // first map is the iterator, second is for the `Option<_>`
+    Ok(array.iter().map(|x| x.map(|x| op(x))).collect())
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/crypto_expressions.rs:91]:
```suggestion
    Ok(array.iter().map(|x| x.map(op)).collect())
```

File: rust/datafusion/src/physical_plan/crypto_expressions.rs [hunk 13]
```diff
@@ -49,58 +58,142 @@ fn sha_process<D: SHA2Digest + Default>(input: &str) -> SHA2DigestOutput<D> {
...
+                other,
+            ))),
+        },
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(a) => {
+                let result = a.as_ref().map(|x| md5_process(x));
+                Ok(ColumnarValue::Scalar(ScalarValue::Utf8(result)))
+            }
+            ScalarValue::LargeUtf8(a) => {
+                let result = a.as_ref().map(|x| md5_process(x));
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/crypto_expressions.rs:170]:
```suggestion
                let result = a.as_ref().map(md5_process);
```

**Comment by maxburke** [comment]
> @andygrove / @Dandandan / @seddonm1 / @maxburke / @jhorstmann -- what are your thoughts on this one?
> 
> It is a significant enough change I think someone using DataFusion in their projects, other than myself should weigh in. I like it a lot -- and I think it could serve as the basis for constant folding (e.g. transform `A + (5 + 7)` --> `A + 12` at plan time)

On the surface this looks like it'll be a fantastic change to have; I am curious to see what the measured impact on query times will be.

(cc @velvia / @mcassels)

File: rust/datafusion/src/physical_plan/crypto_expressions.rs (hunk 12 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/crypto_expressions.rs:91]: Good idea, though, I also though it would work. However, because the functions have different signatures, a deref is needed and thus we need to write it explicitly. Same for `md5_process`.

**Comment by jorgecarleitao** [comment]
Rebased. 💦

**Comment by jorgecarleitao** [comment]
This is now ready to merge. This will collide with an also large PR, #9243. We have a function to enable compatibility, but it is still some work :(

**Comment by alamb** [comment]
@seddonm1  -- what do you think about merge order of this PR and #9243 ? (which will conflict)

**Comment by seddonm1** [comment]
Unfortunately (for me) this logically does go first as being able to identify ScalarValue would give a huge performance advantage. 

I am happy to rework the other one after this is merged.

**Comment by alamb** [comment]
Sounds good -- @jorgecarleitao  let's get it merged ! It looks like it needs another rebase and then we'll get it in

**Comment by velvia** [comment]
Commenting since @maxburke pinged me.

On the surface I think this is a great change from the performance perspective.  I totally agree that being able to deal with scalars instead of just arrays adds huge room for optimization.   I have always thought that always needing intermediate arrays slowed down the processing of DataFusion significantly, for certain cases, and is also cache unfriendly.

On the other hand, I hear what @alamb and others are saying that it adds complexity to what is already nontrivial.   I agree with that.

I wonder if it is possible to get the best of both worlds, by extending the `Array` trait slightly and having a subclass of `Array` which denotes scalars, like `ScalarArray` which do not need `Buffer` storage and just represents constant scalars.   This way, functions would only need to deal with Array, but can recognize this subclass `ScalarArray` and do optimizations that way.   This is a very half-formed thought at the moment.

File: rust/datafusion/src/physical_plan/expressions/nullif.rs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            "{:?} args were supplied but NULLIF takes exactly two args",
+            args.len(),
+        )));
+    }
+
+    let (lhs, rhs) = (&args[0], &args[1]);
+
+    match (lhs, rhs) {
+        (ColumnarValue::Array(lhs), ColumnarValue::Scalar(rhs)) => {
+            let cond_array = binary_array_op_scalar!(lhs, rhs.clone(), eq).unwrap()?;
```
Comment (velvia) [inline rust/datafusion/src/physical_plan/expressions/nullif.rs:86]: This is definitely one of the functions I would expect to benefit the most from this change, given that NULLIF most of the time the right hand argument is likely to be a scalar.

**Comment by seddonm1** [comment]
> I wonder if it is possible to get the best of both worlds, by extending the `Array` trait slightly and having a subclass of `Array` which denotes scalars, like `ScalarArray` which do not need `Buffer` storage and just represents constant scalars. This way, functions would only need to deal with Array, but can recognize this subclass `ScalarArray` and do optimizations that way. This is a very half-formed thought at the moment. The train of thought is just what if the `Array` was not strictly a buffer-based representation but just a way to access columnar data, and in certain cases represents scalars.

I also have many instances where knowing the `ScalarArray` vs `Array` would provide huge performance opportunities in the big PR implement Postgres functions : https://github.com/apache/arrow/pull/9243 - especially for things like Regex matching. It would be relatively trivial to implement if it could be pattern matched.

**Comment by jorgecarleitao** [comment]
@velvia , @seddonm1  My opinion atm is that would not really help much from the compute side, as we would still need to write separate logic for the 4 cases `(Array, scalar)`, `(scalar, Array)`, `(scalar, scalar)`, `(array, array)`, which is exactly what this PR already proposes.

For example, a subtraction of two arrays goes along the lines of

```rust
let iter = lhs.values().zip(rhs.values()).map(|(r, l)| r - l);
let array = unsafe { PrimitiveArray<T>::from_trusted_len_iter(iter) };
// .values() is a slice
```

If we had an implementation of `PrimitiveScalarArray<T>`, we could change `.values()` to be an iterator, but then we lose the benefits of contiguous regions, as Rust can no longer assume that `values` is a contiguous memory region (slices have that property, iterators do not). If both sides are scalars, we want to to create a `PrimitiveScalarArray<T>`, so, again we need to branch. In all cases, we need a match `(lhs, rhs)`. I think that DataFusion in this respect is doing things right. 

To change the Array trait we basically need a change in the arrow specification so that all implementations agree on how to communicate such information via IPC, ffi, etc, so, that is mailing list material :)

Note that this PR addresses @alamb 's concern by introducing a adapter that people can use if they do not want to bother implementing the scalar variants.

**Comment by alamb** [comment]
> Note that this PR addresses @alamb 's concern by introducing a adapter that people can use if they do not want to bother implementing the scalar variants.

Yes, the adapter function I think lowers the barrier to implementing initial scalar functions and should be an easier upgrade path.

TLDR it is that one can call `make_scalar_function(your_old_func_implementation)` and not worry about the updates.

I think this PR is ready, is a great step forward. Let's merge it and continue to iterate.
